Gh-401: Document named operations (#404)
gchq/gaffer-doc: docs/administration-guide/named-operations.md

- In a controlled way, allowing specific Operation Chains to be run by a user that would not normally have permission to run them.

There are [three operations](../reference/operations-guide/named.md) which manage Named Operations. 
These are `AddNamedOperation`, `GetAllNamedOperations` and `DeleteNamedOperations`.

## Using Named Operations

All Named Operations are stored in a cache, so your first step should be to configure a suitable cache.
For details on potential caches and how to configure them, see the [Stores Guide](../administration-guide/gaffer-stores/store-guide.md/#caches).

!!! Note
    If you choose a non-persistent cache then any Named Operations will be lost when you shut down your instance of Gaffer.

This guide assumes that you have set up your graph, if not please refer to our [example deployment](../development-guide/example-deployment/project-setup.md) to get started.

This walkthrough will use this directed graph:

``` mermaid
graph TD
  1(1, count=3) -- count=3 --> 2
  1 -- count=1 --> 4
  2(2, count=1) -- count=2 --> 3
  2 -- count=1 --> 4(4, count=1)
  2 -- count=1 --> 5(5, count=3)
  3(3, count=2) -- count=4 --> 4
```

> Once you have configured your cache, you can then execute your first NamedOperation. 

Reviewer: Maybe I would say something like "Once you have configured your cache, you can then **add** your first NamedOperation". This can then follow onto the next paragraph.

                operations=[ 
                    g.GetAdjacentIds( 
                        include_incoming_out_going="OUTGOING" 
                    ), 
                    g.GetAdjacentIds( 
                        include_incoming_out_going="OUTGOING" 
                    ) 
                ] 
            ), 
            operation_name="2-hop", 
            description="2 hop query", 
            read_access_roles=[ 
                "read-user" 
            ], 
            write_access_roles=[ 
                "write-user" 
            ], 
            overwrite_flag=True 
        )
        ```

Following on from this, you can then run your new NamedOperation:

!!! example "Running your NamedOperation"

    === "Java"

        ``` java
        final NamedOperation<EntityId, CloseableIterable<EntityId>> operation =
        new NamedOperation.Builder<EntityId, CloseableIterable<EntityId>>()
                .name("2-hop")
                .input(new EntitySeed(1))
                .build();
        ```

    === "JSON"

        ``` json
        {
            "class" : "NamedOperation",
            "input" : [ {
                "class" : "EntitySeed",

>                 "class" : "EntitySeed",

Reviewer: Duplicated line

The following code adds a NamedOperation with a parameter that allows the result limit for the OperationChain to be set:

!!! example "Adding a NamedOperation with parameters"

    === "Java"

        ``` java
        final String opChainString = new JSONObject()
            .put("operations", new JSONArray()
                .put(new JSONObject()
                    .put("class", "GetAdjacentIds")
                    .put("includeIncomingOutGoing", "OUTGOING"))
                .put(new JSONObject()
                    .put("class", "GetAdjacentIds")
                    .put("includeIncomingOutGoing", "OUTGOING"))
                .put(new JSONObject()
                    .put("class", "Limit")
                    .put("resultLimit", "${param1}")))
            .toString();

        ParameterDetail param = new ParameterDetail.Builder()
                .defaultValue(1L)
                .description("Limit param")
                .valueClass(Long.class)
                .build();
        Map<String, ParameterDetail> paramMap = Maps.newHashMap();
        paramMap.put("param1", param);

        final AddNamedOperation operation = new AddNamedOperation.Builder()

>                 .operationChain(opChainString)

Reviewer: Does this have to be done using a String or could you give an OperationChain object out of interest?

Author: No it can be done using an OperationChain object if you wanted :) https://github.com/gchq/Gaffer/blob/develop/core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java#L275-L283

                        .then(new GetAdjacentIds.Builder()
                                .inOutType(SeededGraphFilters.IncludeIncomingOutgoingType.OUTGOING)
                                .build())
                        .build())
                .description("2 hop query")
                .name("2-hop")
                .readAccessRoles("read-user")
                .writeAccessRoles("write-user")
                .overwrite()
                .build();
        ```

    === "JSON"

        ``` json
        {
        "class" : "AddNamedOperation",
        "operationName" : "2-hop",
        "description" : "2 hop query",
        "operationChain" : {
            "operations" : [ {
                "class" : "uk.gov.gchq.gaffer.operation.impl.get.GetAdjacentIds",
                "includeIncomingOutGoing" : "OUTGOING"
            }, {
                "class" : "uk.gov.gchq.gaffer.operation.impl.get.GetAdjacentIds",
                "includeIncomingOutGoing" : "OUTGOING"
            } ]
        },
        "overwriteFlag" : true,
        "readAccessRoles" : [ "read-user" ],
        "writeAccessRoles" : [ "write-user" ]
        }
        ```

    === "Python"

        ``` python

>         g.AddNamedOperation( 

Reviewer 2: Quite minor but I think the indentation is a bit off here should really indent by 4 for line continuations inPython

                g.EntitySeed( 
                    vertex=1 
                ) 
            ], 
            operation_name="2-hop" 
        )
        ```

This produces the following result:
```java
EntitySeed[vertex=4]
EntitySeed[vertex=3]
EntitySeed[vertex=5]
```

Named Operations are able to take parameters which allow the OperationChain that is being executed to be configured.
Parameters could be as simple as specifying `resultLimit` on a Limit Operation, they could specify a custom view to use in an Operation or even the input to an Operation.

When adding a NamedOperation with parameters to an OperationChain it must be specified as a JSON string, with the parameter names enclosed in `${` and `}`. 
For each parameter, a `ParameterDetail` object must be created which gives a description, a class type and an optional default for that parameter.
As the default is optional you can alternatively indicate that the parameter must be provided and that there is no default.

The following code adds a NamedOperation with a parameter that allows the result limit for the OperationChain to be set:

!!! example "Adding a NamedOperation with parameters"

    === "Java"

        ``` java

>         final String opChainString = "{" +

Reviewer 2: Not sure if its worth it but this seems like it could be very error prone to typos would using a `JSONObject()` and building it up be a better example e.g. something like:

```java
String opChainString = new JSONObject()
    .put("operations", new JSONArray()
        .put(new JSONObject()
            .put("class", "GetAdjacentIds")
            .put("includeIncomingOutGoing", "OUTGOING"))
        .put(new JSONObject()
            .put("class", "GetAdjacentIds")
            .put("includeIncomingOutGoing", "OUTGOING"))
        .put(new JSONObject()
            .put("class", "Limit")
            .put("resultLimit", "${param1}")))
    .toString();
```

Reviewer 3: I agree that this would be better.

# Named Operations

> This guide walks you through how to configure Gaffer to allow you to execute Named Operations.

Reviewer 3: ```suggestion
This guide walks through configuring Gaffer to use Named Operations and how to run them.
```

Named Operations allow users to encapsulate an OperationChain into a new single NamedOperation.
When this NamedOperation is executed, just like any other Operation, it will execute the encapsulated OperationChain.

> Named Operations can then be added to Operation Chains and executed as you would any other Operation.

Reviewer 3: ```suggestion
Named Operations execute the encapsulated Operation Chain and are used just like any other Operation.
```

There are various possible uses for Named Operations:

- Making it simpler to store and run frequently used Operation Chains.
- In a controlled way, allowing specific Operation Chains to be run by a user that would not normally have permission to run them.

> There are [several operations](../reference/operations-guide/named.md) which manage Named Operations. 

Reviewer 3: ```suggestion
There are [three operations](../reference/operations-guide/named.md) for managing Named Operations. 
```

## Creating Named Operations

Any Named Operations you create will be stored in a cache, so your first step should be to configure a suitable cache.
For details on potential caches and how to configure them, see the [Stores Guide](../administration-guide/gaffer-stores/store-guide.md/#caches).

!!! Note
    If you choose a non-persistent cache then any Named Operations will be lost when you shut down your instance of Gaffer.

This walkthrough will use this directed graph:

``` mermaid
graph TD
  1(1, count=3) -- count=3 --> 2
  1 -- count=1 --> 4
  2(2, count=1) -- count=2 --> 3
  2 -- count=1 --> 4(4, count=1)
  2 -- count=1 --> 5(5, count=3)
  3(3, count=2) -- count=4 --> 4
```

Once you have configured your cache, you can then create your first NamedOperation. 
You should start by creating your user instance and graph:

``` java
// Create your user
final User user = new User("user01");

// Build your graph instance
final Graph graph = new Graph.Builder()
        .config(getDefaultGraphConfig())
        .addSchemas(StreamUtil.openStreams(getClass(), schemaPath))
        .storeProperties(getDefaultStoreProperties())
        .build();

> ```

Reviewer 3: I think it would be better to link to an administration page which covers setting up a graph. This example from the old docs is not something which can be used as is and I don't think it adds anything.

        ``` json
        {
            "class" : "NamedOperation",
            "input" : [ {
                "class" : "EntitySeed",
                "class" : "EntitySeed",
                "vertex" : 1
            } ],
            "operationName" : "2-hop-with-limit",
            "parameters" : {
                "param1" : 2
            }
        }
        ```

    === "Python"

        ``` python
        g.NamedOperation( 
            input=[ 
                g.EntitySeed( 
                    vertex=1 
                ) 
            ], 
            operation_name="2-hop-with-limit", 
            parameters={'param1': 2} 
        )
        ```

This will produce these results:

```java
EntitySeed[vertex=4]
EntitySeed[vertex=3]
```

Details of all available NamedOperations can be fetched using the `GetAllNamedOperations` operation:

!!! example "Fetching Named Operations"

    === "Java"

        ``` java
        final GetAllNamedOperations operation = new GetAllNamedOperations();
        ```

    === "JSON"

        ``` json
        {
            "class" : "GetAllNamedOperations"
        }
        ```

    === "Python"

        ``` python
            g.GetAllNamedOperations()   

>         ```

Reviewer 3: This duplicates the example on the reference guide page. It can be removed.

> For more examples of Named Operations, please refer to the Reference Guide on [Operations](../reference/operations-guide/named.md).

Reviewer 3: ```suggestion
For more examples of Named Operations, please refer to the [Named Operations page in the Reference Guide](../reference/operations-guide/named.md).
```
